### ISSUE TITLE
False-negative type errors using async functions

### ISSUE THREAD

**Issue by doctyper** [codemix/flow-runtime#67]
This is a:

 - [x] Bug Report
 - [ ] Feature Request
 - [ ] Question
 - [ ] Other

Which concerns:

 - [x] flow-runtime
 - [x] babel-plugin-flow-runtime
 - [ ] flow-runtime-validators
 - [ ] flow-runtime-mobx
 - [ ] flow-config-parser
 - [ ] The documentation website

---

### What is the current behaviour?

It appears flow-runtime gets confused type checking async functions. The following gives me an error in https://codemix.github.io/flow-runtime/#/try (default options):

```js
const bar = async ({
    id
}: {
    id: string;
}) => {
    return await new Promise(resolve => setTimeout(() => resolve(id), 1000));
};

const foo = (id: string) => {
    return bar({
        id
    })
  			.then(id => console.log(id))
        .catch((e: Error) => console.error("foo", e));
};

foo("testId");
```

results in:

```shell
foo TypeError: arguments[0].id must be a string

Expected: string

Actual: void


    at _callee$ (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:20:169)
    at tryCatch (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:59153:41)
    at Generator.invoke [as _invoke] (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:59444:23)
    at Generator.prototype.(anonymous function) [as next] (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:59205:22)
    at step (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:11:191)
    at eval (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:11:437)
    at eval (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:11:99)
    at bar (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:41:21)
    at foo (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:52:12)
    at eval (eval at run (https://codemix.github.io/flow-runtime/static/js/main.1ab3461d.js:20454:17), <anonymous>:65:1)
```

Whereas non-async functions behave as expected:

```js
const bar = ({
    id
}: {
    id: string;
}) => {
    return new Promise(resolve => setTimeout(() => resolve(id), 1000));
};

const foo = (id: string) => {
    return bar({
        id
    })
  			.then(id => console.log(id))
        .catch((e: Error) => console.error("foo", e));
};

foo("testId");
```

results in:

```
testId
```

---

### What is the expected behaviour?

It should not display false-negative results.

---

### Which package versions are you using?

Node 6.9.1

flow-runtime@0.7.0
babel-plugin-flow-runtime@0.7.0

**Comment by phpnode** [comment]
ok, this happens because the hack we use to maintain reference to `arguments` in an arrow function (which we need for destructuring but is not ordinarily allowed) is getting mangled by regenerator. Curiously though, the official babel hack for this using `arg._shadowedFunctionLiteral = path` does not appear to work - `arguments` gets hoisted up too high. 

TL;DR - To fix this we need to create temporary vars to hold arguments and roll the actual destructuring into the function body, which is really a bit of a pain.

**Comment by phpnode** [comment]
released in v0.8.0, this fix has caused a bit of a regression for error messages when you use destructuring, you'll likely get `Value is ...` rather than `arguments[0] is ...`. I'd like to overhaul how this works generally so will revisit this then.
